Logintests

AndreaGonzalezR opened the pull request.

AndreaGonzalezR committed:
Workspace Setup

AndreaGonzalezR committed:
wdio.conf.js was added to the workspace setup

AndreaGonzalezR committed:
Test cases explanation and Login tests were made

AndreaGonzalezR committed:
Changes in the values of some fields related to package.json were made

AndreaGonzalezR committed:
Some changes in this document were introduced

AndreaGonzalezR committed:
Some changes were introduce in order to run the tests successfuly

AndreaGonzalezR committed:
This document was created

AndreaGonzalezR committed:
This document was created when trying to save the workspace as an archive with .code-workspace extension

AndreaGonzalezR committed:
Test cases with their assertions were successfully created

AndreaGonzalezR committed:
Login Tests README.MD updated

Instructions in order to run the Login Tests were added

rquintana-abstracta commented on README.md:
```diff
@@ -1 +1,7 @@
\ No newline at end of file
+# Digital Bank 
+
+Para correr los test es necesario acceder al espacio de trabajo en Visual Studio Code, denominado E-BANKING WORKSPACE. 
```
We defined the convention of **everything in english**, including documentation, class, functions and variable names, etc. Make sure everything is in english in this repo

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,91 @@
+   
+    it('Login  with correct username and correct password', async () => {
+
+        var expect = require('chai').expect;
```
**expect()** is globally defined in newer versions of webdriverio, there's no need to explicitly require it. Delete this line

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,91 @@
+    
+    it ('Log in with incorrect username and correct password', async () => {
+
+        var expect = require('chai').expect;
```
Same comment as above regarding expect being defined globally

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,91 @@
+describe ('TestLogin', () => {
+
+   
+    it('Login  with correct username and correct password', async () => {
```
Consider using the naming convention of structuring your test case's names with a **should** based sentence. In your case, something like: "**Should** login with...". Same with every other test.

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,91 @@
+
+        await browser.url('');
+        
+        let username = await $('[id = "username"]');
```
Use **page objects** to abstract out logic regarding elements location / interaction, then make the tests interact with pages exclusively through them. That way you decouple low-level details, such as locating elements, calling the right webdrvierio method on them, etc, from actual test definition.

Ideally the flow will resamble something like this, after you create the loginPage page object:

```javascript
loginPage.open('login');
loginPage.logIn('jsmith@demo.io', 'Demo123!');
```

Basically encapsulating the complexities of loging in into methods with higher-level names

See: https://webdriver.io/docs/pageobjects/

rquintana-abstracta commented on wdio.conf.js:
```diff
@@ -0,0 +1,293 @@
+    // with `/`, the base url gets prepended, not including the path portion of your baseUrl.
+    // If your `url` parameter starts without a scheme or `/` (like `some/path`), the base url
+    // gets prepended directly.
+    baseUrl: 'http://localhost:8080/bank/login',
```
This should be the base URL por the whole application, as this scope affects every test. BaseUrl in this context should be `http://localhost:8080/bank/`, and every other test that needs access to a specific location should only need to append the string that goes after: like `login` or `home`

rquintana-abstracta reviewed (changes requested):
Implement requested changes

AndreaGonzalezR committed:
Page object Pattern applied to Login Page tests

AndreaGonzalezR committed:
New test cases added and fundamentation written in English

AndreaGonzalezR committed:
Update README.md

AndreaGonzalezR committed:
Changes in the baseUrl  in order to access correctly to the web site

AndreaGonzalezR committed:
Merge branch 'logintests' of https://github.com/AndreaGonzalezR/e-banking into logintests

# Conflicts:
#	README.md

rquintana-abstracta commented on E-BANKING WORKSPACE.code-workspace:
```diff
@@ -0,0 +1,7 @@
+{
```
This file is specific for your setup, add it to gitignore, as it's not relevant for the project

AndreaGonzalezR committed:
E-BANKING WORKSPACE.code-workspace was changed to gitignore

rquintana-abstracta commented on pages/base.page.js:
```diff
@@ -0,0 +1,44 @@
+    async emptyFieldAndSendText(element, text){
+
+        await element.waitForClickable({ timeout: PAGE_TIMEOUT });
+        await element.clearValue();
+        await element.click();
+        await element.keys(text);
```
If you use the `setValue()` method wdio provides for web elements, you can avoid much of this logic, simplifying method to just:

```javascript
async emptyFieldAndSendText(element, text){
     await element.waitForClickable({ timeout: PAGE_TIMEOUT });
     await element.setValue(text);
}
```

rquintana-abstracta commented on pages/home.page.js:
```diff
@@ -0,0 +1,15 @@
+
+    get welcomeUser(){return $('.active')}
+
+    async getWelcomeUserText(){
```
Remove this method, as it's not really needed for validating the text of a welcome user (you can use a wdio specific matcher to retrieve it given a web element)

rquintana-abstracta commented on pages/login.page.js:
```diff
@@ -0,0 +1,29 @@
+
+    }
+
+    async selectRememberMe(userName, passWord){
```
This method is duplicating much of the logic defined in `loginAs()`.

Consider narrowing the scope of it by only assigning it the responsibility of selecting the remember me option (as its name imply):

```javascript
async selectRememberMe(){
   await this.rememberMe.click();
``` 

And, if you need to login as well as part of a test, just call your current `loginAs()` method after calling the `selectRememberMe()` method. That way you'll login with that option selected, and you won't have to define the logic twice.

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,79 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('andrea@demo.io', 'Demo123!');
+        expect(await $('badge badge-pill badge-danger')).toHaveTextContaining('Error');
```
By using the page object pattern, it implies the tests won't interact with raw elements directly. You are mixing levels of abstraction here: Page object methods with raw element selectors. Represent this error message element as an attribute from the LoginPage class, like you did with `usernameInput`, and use it for this assertion, so you'll end up with something like:

```javascript
expect(await LoginPage.statusMessage).toHaveTextContaining('Error');
```

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,79 @@
+       
+        await LoginPage.open('login');
+        await LoginPage.loginAs('jsmith@demo.io', 'Demo');
+        expect(await $('badge badge-pill badge-danger')).toHaveTextContaining('Error');
```
Same comment about encapsualting that badge element into an attribute from `LoginPage`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,79 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('', '');
+        expect(await $('badge badge-pill badge-danger')).toHaveTextContaining('Error');
```
Same comment about encapsualting that badge element into an attribute from `LoginPage`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,79 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('', 'Demo123!');
+        expect(await $('badge badge-pill badge-danger')).toHaveTextContaining('Error');
```
Same comment about encapsualting that badge element into an attribute from `LoginPage`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,79 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('jsmith@demo.io', '');
+        expect(await $('badge badge-pill badge-danger')).toHaveTextContaining('Error');
```
Same comment about encapsualting that badge element into an attribute from `LoginPage`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,79 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('', 'Demo');
+        expect(await $('badge badge-pill badge-danger')).toHaveTextContaining('Error');
```
Same comment about encapsualting that badge element into an attribute from `LoginPage`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,79 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('andrea@demo.io', '');
+        expect(await $('badge badge-pill badge-danger')).toHaveTextContaining('Error');
```
Same comment about encapsualting that badge element into an attribute from `LoginPage`

rquintana-abstracta commented on tests/testlogout.js:
```diff
@@ -0,0 +1,31 @@
+//const { default: $ } = require('webdriverio/build/commands/browser/$');
```
Remove this file, as it's not part of the Login related logic, which is the focus of this Pull Request

rquintana-abstracta commented on README.md:
```diff
@@ -1 +1,10 @@
-# e-banking
\ No newline at end of file
+# App Digital Bank 
```
This description should assume the user does not necessarily have the tools necessary for running the tests. Consider adding a pre-requisites section that mentions the need for having node.js installed in their system.

Also, they need to install the dependencies stated in your package.json before being able to run the tests: (That's achieved by running: `npm install` on your project's root directory). Only after that they'll be able to run the tests.

rquintana-abstracta commented on pages/base.page.js:
```diff
@@ -0,0 +1,44 @@
+export default class BasePage{
+    
+    /**
+     * Abrir pagina
```
These comments need to be in english as well. Translate them on every element you have them.

rquintana-abstracta reviewed (changes requested):
Implement requested changes

AndreaGonzalezR committed:
testlogout.js was removed and testlogin.js code was improved

testlogin.js was modified to encapsulate the badge element into an attribute and to improve the selectRememberMe() as well. testlogout.js was deleted because it is not part of the Login related logic

AndreaGonzalezR committed:
Improvements to Page Object Pattern

The pattern Page Object was modified in order to make the code more efficient

AndreaGonzalezR committed:
Update README.md

AndreaGonzalezR committed:
logout file was created

AndreaGonzalezR committed:
Revert "logout file was created"

This reverts commit 8d93cd542204858d402708eb1b53fafef2efae8b.

AndreaGonzalezR committed:
Revert "Revert "logout file was created""

This reverts commit f271128ae9622130e084d4bc0a17be79943babb1.

rquintana-abstracta commented on pages/login.page.js:
```diff
@@ -0,0 +1,26 @@
+    get passWordInput(){return $('#password')}
+    get loginButton(){return $('#submit')}
+    get rememberMe(){return $('#remember-me')}
+    get statusMessage(){return $('#badge badge-pill badge-danger')}
```
You don't need the whole chain of badges to select this. Prefer only `.badge` (notice I used a class selector instead of an id based, as you used, because badge is a class in this case, so `#badge` does not exist in the page), as the rest of them capture style specific stuff that you don't need for individualizing the element.

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,80 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('andrea@demo.io', 'Demo123!');
+        expect(await LoginPage.statusMessage).toHaveTextContaining('Error');
```
In order for this to work, `await` needs to be paced at the begining of the line, instead of before LoginPage. If not, it gives false positives as we discovered during first workshop. Place it there instead.

```javascript
await expect(LoginPage.statusMessage).toHaveTextContaining('Error');

```

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,80 @@
+       
+        await LoginPage.open('login');
+        await LoginPage.loginAs('jsmith@demo.io', 'Demo');
+        expect(await LoginPage.statusMessage).toHaveTextContaining('Error');
```
Same comment about the placing of `await`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,80 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('', '');
+        expect(await LoginPage.statusMessage).toHaveTextContaining('Error');
```
Same comment about the placing of `await`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,80 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('', 'Demo123!');
+        expect(await LoginPage.statusMessage).toHaveTextContaining('Error');
```
Same comment about the placing of `await`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,80 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('jsmith@demo.io', '');
+        expect(await LoginPage.statusMessage).toHaveTextContaining('Error');
```
Same comment about the placing of `await`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,80 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('', 'Demo');
+        expect(await LoginPage.statusMessage).toHaveTextContaining('Error');
```
Same comment about the placing of `await`

rquintana-abstracta commented on tests/testlogin.js:
```diff
@@ -0,0 +1,80 @@
+
+        await LoginPage.open('login');
+        await LoginPage.loginAs('andrea@demo.io', '');
+        expect(await LoginPage.statusMessage).toHaveTextContaining('Error');
```
Same comment about the placing of `await`

rquintana-abstracta reviewed (changes requested):
Implement requested changes

AndreaGonzalezR committed:
statusMessage selector was changed to a class, inorder to improve it

AndreaGonzalezR committed:
testlogout.js was deleted and testlogin.js expects were improved

AndreaGonzalezR committed:
Merge branch 'logintests' of https://github.com/AndreaGonzalezR/e-banking into logintests
